File: liblouisutdml/liblouisutdml.c
/* liblouisutdml Braille Transcription Library

   This file may contain code borrowed from the Linux screenreader
   BRLTTY, copyright (C) 1999-2006 by
   the BRLTTY Team

   Copyright (C) 2004, 2005, 2006
   ViewPlus Technologies, Inc. www.viewplus.com
   and
   JJB Software, Inc. www.jjb-software.com
   All rights reserved

   This file is free software; you can redistribute it and/or modify it
   under the terms of the Lesser or Library GNU General Public License 
   as published by the
   Free Software Foundation; either version 3, or (at your option) any
   later version.

   This file is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
   Library GNU General Public License for more details.

   You should have received a copy of the Library GNU General Public 
   License along with this program; see the file COPYING.  If not, write to
   the Free Software Foundation, 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301, USA.

   Maintained by John J. Boyer john.boyer@jjb-software.com
   */

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "louisutdml.h"
#include <libxml/HTMLparser.h>
#include <libxml/catalog.h>

UserData *ud = NULL;

LBUAPI char *EXPORT_CALL
lbu_version ()
{
  static char *version = PACKAGE_VERSION;
  static char bothVersions[60];
  strcpy (bothVersions, version);
  strcat (bothVersions, " ");
  strcat (bothVersions, lou_version ());
  return bothVersions;
}

LBUAPI void EXPORT_CALL lbu_loadXMLCatalog(const char *filename)
{
  xmlLoadCatalog(filename);
}
void
libxml_errors (void *ctx ATTRIBUTE_UNUSED, const char *msg, ...)
{
  va_list args;
  char buffer[MAXNAMELEN];
  va_start (args, msg);
  memset (buffer, 0, sizeof (buffer));
  vsnprintf (buffer, sizeof (buffer) - 4, msg, args);
  va_end (args);
  logMessage (LOU_LOG_ERROR, "%s", buffer);
}

static xmlParserCtxt *ctxt;
static int libxml2_initialized = 0;

static void
cleanupLibxml ()
{
  destroy_semantic_table ();
  if (ud->doc != NULL)
    xmlFreeDoc (ud->doc);
  if (!libxml2_initialized)
    return;
  xmlCleanupParser ();
  initGenericErrorDefaultFunc (NULL);
  xmlFreeParserCtxt (ctxt);
}

static void
freeEverything ()
{
  lbu_logEnd ();
  cleanupLibxml ();
  lbu_free ();
}

static int
processXmlDocument (const char *inputDoc, int length)
{
  /*This function does all processing of xml documents as such.
   * If length is 0 the document is assumed to be a file.
   * If length is not 0 it is assumed to be in memory.
   * Sort of hackish, but only hackers will see it. */
  xmlNode *rootElement = NULL;
  int haveSemanticFile;
  if (!libxml2_initialized)
    {
      libxml2_initialized = 1;
      LIBXML_TEST_VERSION xmlKeepBlanksDefault (0);
      xmlSubstituteEntitiesDefault (1);
      xmlThrDefIndentTreeOutput (1);
      xmlThrDefKeepBlanksDefaultValue (0);
      xmlThrDefLineNumbersDefaultValue (1);
    }
  ud->doc = NULL;
  ctxt = xmlNewParserCtxt ();
  xmlSetGenericErrorFunc (ctxt, libxml_errors);
  if (length == 0)
    {
      if ((ud->mode & htmlDoc))
	ud->doc = htmlParseFile (inputDoc, NULL);
      else
	{
	  if (ud->internet_access)
	    ud->doc = xmlCtxtReadFile (ctxt, inputDoc, NULL,
				       XML_PARSE_DTDVALID | XML_PARSE_NOENT);
	  else
	    ud->doc = xmlParseFile (inputDoc);
	  if (ud->doc == NULL)
	    {
	      logMessage
		(LOU_LOG_FATAL, "Document could not be processed; may be \
malformed or contain illegal characters");
	      cleanupLibxml ();
	      return 0;
	    }
	}
    }
  else
    ud->doc = xmlParseMemory (inputDoc, length);
  if (ud->doc == NULL)
    {
      logMessage (LOU_LOG_FATAL, "Document could not be processed, probably  malformed");
      cleanupLibxml ();
      return 0;
    }
  if (ud->doc->encoding == NULL)
    {
      logMessage (LOU_LOG_ERROR, "Encoding, preferably UTF-8,  must be specified");
      cleanupLibxml ();
      return 0;
    }
  if (ud->format_for >= utd && ignore_case_comp (ud->doc->encoding,
						 "UTF-8", 5) != 0)
    {
      logMessage (LOU_LOG_ERROR, "UTDML requires UTF-8 encoding, not '%s'",
		    ud->doc->encoding);
      cleanupLibxml ();
      return 0;
    }
  rootElement = xmlDocGetRootElement (ud->doc);
  if (rootElement == NULL)
    {
      logMessage (LOU_LOG_ERROR, "Document is empty");
      cleanupLibxml ();
      return 0;
    }
  if (ud->mode & convertOnly)
    convert_utd ();
  else
    {
      haveSemanticFile = compile_semantic_table (rootElement);
      do_xpath_expr ();
      examine_document (rootElement);
      append_new_entries ();
      /* This will generate a new semantic-action file if none exists 
       * and newEntries is  yes in the configuration file.
       * Otherwise it generates a new_enhtries file.*/
      if (!haveSemanticFile)
	{
	  cleanupLibxml ();
	  return 0;
	}
      if (!transcribe_document (rootElement))
	{
	  logMessage (LOU_LOG_ERROR, "Document could not be transcribed");
	  cleanupLibxml ();
	  return 0;
	}
    }
  cleanupLibxml ();
  return 1;
}

LBUAPI void *EXPORT_CALL
lbu_initialize (const char *configFileList,
		const char *logFileName, const char *settingsString)
{
  if (!read_configuration_file (configFileList, logFileName,
				settingsString, 0))
    return NULL;
  return (void *) ud;
}

LBUAPI int EXPORT_CALL
lbu_translateString (const char *configFileList,
		     const char *inbuf, int inlen, widechar * outbuf,
		     int *outlen,
		     const char *logFileName, const char *settingsString,
		     unsigned int mode)
{
/* Translate the well-formed xml expression in inbuf into braille 
* according to the specifications in configFileList. If the expression 
* is not well-formed or there are oteer errors, return 0.*/
  int k;
  char *xmlInbuf;
  if (!read_configuration_file
      (configFileList, logFileName, settingsString, mode))
    return 0;
  if (inbuf == NULL || outbuf == NULL || outlen == NULL)
    return 0;
  ud->inbuf = inbuf;
  ud->inlen = inlen;
  ud->outbuf = outbuf;
  ud->outlen = *outlen;
  ud->inFile = ud->outFile = NULL;
  for (k = 0; k < inlen; k++)
    if (inbuf[k] > ' ')
      break;
  if (inbuf[k] != '<')
    {
      if (ud->format_for == utd)
	k = utd_transcribe_text_string ();
      else
	k = transcribe_text_string ();
      *outlen = ud->outlen_so_far;
      lbu_logEnd ();
      return k;
    }
  if (inbuf[k + 1] == '?')
    xmlInbuf = (char *) inbuf;
  else
    {
      inlen += strlen (ud->xml_header);
      if (!(xmlInbuf = malloc (inlen + 4)))
	{
	  logMessage (LOU_LOG_FATAL, "Not enough memory");
	  return 0;
	}
      strcpy (xmlInbuf, ud->xml_header);
      strcat (xmlInbuf, "\n");
      strcat (xmlInbuf, inbuf);
    }
  k = processXmlDocument (xmlInbuf, inlen);
  *outlen = ud->outlen_so_far;
  if (xmlInbuf != inbuf)
    free (xmlInbuf);
  lbu_logEnd ();
  return k;
}

LBUAPI int EXPORT_CALL lbu_translateFile
  (const char *configFileList, const char *inFileName,
   const char *outFileName, const char *logFileName,
   const char *settingsString, unsigned int mode)
{
/* Translate the well-formed xml expression in inFileName into 
* braille according to the specifications in configFileList. If the 
* expression is not well-formed or there are other errors, 
* return 0. */
  int k;
  if (!read_configuration_file
      (configFileList, logFileName, settingsString, mode))
    return 0;
  if (inFileName == NULL || outFileName == NULL)
    return 0;
  if (strcmp (outFileName, "stdout"))
    {
      if (!(ud->outFile = fopen (outFileName, "wb")))
	{
	  logMessage (LOU_LOG_ERROR, "Can't open output file %s.", outFileName);
	  return 0;
	}
    }
  else
    ud->outFile = stdout;
  k = processXmlDocument (inFileName, 0);
  if (ud->outFile != stdout)
    fclose (ud->outFile);
  lbu_logEnd ();
  return k;
}

LBUAPI int EXPORT_CALL lbu_translateTextFile
  (const char *configFileList, const char *inFileName,
   const char *outFileName, const char *logFileName,
   const char *settingsString, unsigned int mode)
{
/* Translate the text file in inFileName into braille according to
* the specifications in configFileList. If there are errors, print 
* an error message and return 0.*/
  int k;
  widechar outbuf[2 * BUFSIZE];
  if (!read_configuration_file
      (configFileList, logFileName, settingsString, mode))
    return 0;
  if (inFileName == NULL || outFileName == NULL)
    return 0;
  ud->outbuf = outbuf;
  ud->outlen = (sizeof (outbuf) / CHARSIZE) - 4;
  if (strcmp (inFileName, "stdin"))
    {
      if (!(ud->inFile = fopen (inFileName, "rb")))
	{
	  logMessage (LOU_LOG_ERROR, "Can't open input file %s.", inFileName);
	  return 0;
	}
    }
  else
    ud->inFile = stdin;
  if (strcmp (outFileName, "stdout"))
    {
      if (!(ud->outFile = fopen (outFileName, "wb")))
	{
	  logMessage (LOU_LOG_ERROR, "Can't open output file %s.", outFileName);
	  return 0;
	}
    }
  else
    ud->outFile = stdout;
  if (ud->format_for == utd)
    k = utd_transcribe_text_file ();
  else
    k = transcribe_text_file ();
  if (!k)
    {
      freeEverything ();
      return 0;
    }
  if (ud->inFile != stdin)
    fclose (ud->inFile);
  if (ud->outFile != stdout)
    fclose (ud->outFile);
  lbu_logEnd ();
  return k;
}

LBUAPI int EXPORT_CALL
lbu_backTranslateString (const char *configFileList,
			 const char *inbuf, int inlen, widechar
			 * outbuf,
			 int *outlen,
			 const char *logFileName, const char
			 *settingsString, unsigned int mode)
{
  int k;
  logMessage(LOU_LOG_INFO, "Begin lbu_backTranslateString: inbuf=%s", inbuf);
  if (!read_configuration_file
      (configFileList, logFileName, settingsString, mode))
    return 0;
  if (inbuf == NULL || outbuf == NULL || outlen == NULL)
    return 0;
  ud->inbuf = inbuf;
  ud->inlen = inlen;
  ud->outbuf = outbuf;
  ud->outlen = *outlen;
  ud->inFile = ud->outFile = NULL;
  if (ud->format_for == utd)
  {
    logMessage(LOU_LOG_DEBUG, "ud->format_for=utd");
    k = utd_back_translate_braille_string ();
  }
  else
    k = back_translate_braille_string ();
  if (!k)
    {
      freeEverything ();
      return 0;
    }
  *outlen = ud->outlen_so_far;
  logMessage(LOU_LOG_INFO, "Finish lbu_backTranslateString");
  lbu_logEnd ();
  return 1;
}

LBUAPI int EXPORT_CALL lbu_backTranslateFile
  (const char *configFileList, const char *inFileName,
   const char *outFileName, const char *logFileName,
   const char *settingsString, unsigned int mode)
{
/* Back translate the braille file in inFileName into either an 
* xml file or a text file according to
* the specifications in configFileList. If there are errors, print an 
* error message and return 0.*/
  int k;
  widechar outbuf[2 * BUFSIZE];
  if (!read_configuration_file
      (configFileList, logFileName, settingsString, mode))
    return 0;
  if (inFileName == NULL || outFileName == NULL)
    return 0;
  ud->outbuf = outbuf;
  ud->outlen = (sizeof (outbuf) / CHARSIZE) - 4;
  if (strcmp (inFileName, "stdin"))
    {
      if (!(ud->inFile = fopen (inFileName, "rb")))
	{
	  logMessage (LOU_LOG_ERROR, "Can't open input file %s.", inFileName);
	  return 0;
	}
    }
  else
    ud->inFile = stdin;
  if (strcmp (outFileName, "stdout"))
    {
      if (!(ud->outFile = fopen (outFileName, "wb")))
	{
	  logMessage (LOU_LOG_ERROR, "Can't open output file %s.", outFileName);
	  return 0;
	}
    }
  else
    ud->outFile = stdout;
  if (ud->format_for == utd)
    k = utd_back_translate_file ();
  else
    k = back_translate_file ();
  if (!k)
    {
      freeEverything ();
      return 0;
    }
  if (ud->inFile != stdin)
    fclose (ud->inFile);
  if (ud->outFile != stdout)
    fclose (ud->outFile);
  lbu_logEnd ();
  return 1;
}

static char writeablePath[MAXNAMELEN];
static char *writeablePathPtr = NULL;

LBUAPI char *EXPORT_CALL
lbu_setWriteablePath (const char *path)
{
  writeablePathPtr = NULL;
  if (path == NULL)
    return NULL;
  strcpy (writeablePath, path);
  writeablePathPtr = writeablePath;
  return writeablePathPtr;
}

LBUAPI char *EXPORT_CALL
lbu_getWriteablePath ()
{
  return writeablePathPtr;
}

LBUAPI int EXPORT_CALL
lbu_charToDots (const char *tableList, const unsigned char *inbuf,
		unsigned char *outbuf, int length, const char *logFile,
		unsigned int mode)
{
  widechar *interBuf;
  int wcLength;
  int utf8Length;
  int result = 0;
  if (tableList == NULL || inbuf == NULL || outbuf == NULL)
    return 0;
  lbu_logFile (logFile);
  interBuf = malloc (length * CHARSIZE);
  utf8Length = length;
  wcLength = length;
  utf8_string_to_wc (inbuf, &utf8Length, interBuf, &wcLength);
  result = lou_charToDots (tableList, interBuf, interBuf, wcLength, mode
			   | ucBrl);
  if (result)
    {
      utf8Length = length;
      wc_string_to_utf8 (interBuf, &wcLength, outbuf, &utf8Length);
    }
  lbu_logEnd ();
  free (interBuf);
  return result;
}

LBUAPI int EXPORT_CALL
lbu_dotsToChar (const char *tableList, const unsigned char *inbuf,
		unsigned char *outbuf, int length, const char *logFile,
		unsigned int mode)
{
  widechar *interBuf;
  int wcLength;
  int utf8Length;
  int result = 0;
  if (tableList == NULL || inbuf == NULL || outbuf == NULL)
    return 0;
  lbu_logFile (logFile);
  interBuf = malloc (length * CHARSIZE);
  utf8Length = length;
  wcLength = length;
  utf8_string_to_wc (inbuf, &utf8Length, interBuf, &wcLength);
  result = lou_dotsToChar (tableList, interBuf, interBuf, wcLength, mode);
  if (result)
    {
      utf8Length = length;
      wc_string_to_utf8 (interBuf, &wcLength, outbuf, &utf8Length);
    }
  lbu_logEnd ();
  free (interBuf);
  return result;
}

LBUAPI int EXPORT_CALL
lbu_checkTable (const char *tableList, const char *logFile, unsigned int mode)
{
  int result = 1;
  lbu_logFile (logFile);
  if (!lou_getTable (tableList))
    result = 0;
  lbu_logEnd ();
  return result;
}

LBUAPI void EXPORT_CALL
lbu_free ()
{
/* Free all memory used by liblouisutdml. You MUST call this function at 
* the END of your application.*/
  lbu_logEnd ();
  lou_free ();
  destroy_semantic_table ();
  if (ud != NULL)
    free (ud);
  ud = NULL;
}

File: java/src/org/liblouis/LibLouisUTDML.java
/* liblouisutdml Braille Transcription Library

   This file may contain code borrowed from the Linux screenreader
   BRLTTY, copyright (C) 1999-2006 by
   the BRLTTY Team

   Copyright (C) 2004, 2005, 2006
   ViewPlus Technologies, Inc. www.viewplus.com
   and
   Abilitiessoft, Inc. www.abilitiessoft.org
   All rights reserved

   This file is free software; you can redistribute it and/or modify it
   under the terms of the Lesser or Library GNU General Public License
   as published by the
   Free Software Foundation; either version 3, or (at your option) any
   later version.

   This file is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   Library GNU General Public License for more details.

   You should have received a copy of the Library GNU General Public
   License along with this program; see the file COPYING.  If not, write to
   the Free Software Foundation, 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301, USA.

   Maintained by John J. Boyer john.boyer@abilitiessoft.org
   */

package org.liblouis;

import java.io.File;

public final class LibLouisUTDML {
/** 
* Bindings for the liblouisutdml and liblouis C libraries.
* These bindings include enough functionality so that it should never 
* be necessary to call liblouis directly. This saves the hassle of 
* dealing with two sets of bindings. liblouisutdml can produce a variety 
* of output types according to the value of the formatFor configuration 
* setting. To get UTDML formatFor is set to utd.
*/

/** 
* Definitions of mode bits 
*
* The bits for liblouis are included because it is sometimes necessary 
* to pass a liblouis mode to a liblouisutdml function. The liblouis bits 
* take up the low-order bits of the mode integer, while the 
* liblouisutdml bits take up the high-order bits, except that the 
* sign bit is skipped.
*/
public static final int NoContractions = 1;
public static final int CompbrlAtCursor = 1<<1;
public static final int DotsIO = 1<<2;
public static final int Comp8Dots = 1<<3;
public static final int Pass1Only = 1<<4;
public static final int CompbrlLeftCursor = 1<<5;
public static final int OtherTrans = 1<<6;
public static final int UcBrl = 1<<7;
public static final int DontInit = 1<<30;
public static final int HtmlDoc = 1<<29;
public static final int notUC = 1<<28;

/**
* This class is a singleton.
*/

private static LibLouisUTDML singleInstance = new LibLouisUTDML();
private static boolean libraryLoaded = false;

  private LibLouisUTDML () {
  }
  
  public static LibLouisUTDML getInstance()
  {
    return singleInstance;
  }
  
  public static native void initialize (String dataPath, String 
  writeablePath, String logFile);
  
  public static void loadLibrary (String libraryPath, String 
  librarySuffix) throws Exception {
    if (libraryPath == null || librarySuffix == null)
    {
      System.out.println("prefix or suffix is null");
      throw new Exception (
      "Could not load libraries. libraryPath or librarySuffix undefined.");
    }
    LibLouis.loadLibrary(libraryPath, librarySuffix);
  }
  
  
  /** Return a string giving the versions of both liblouisutdml and 
   * liblouis */
  public native String version ();

  /**
   * Load a XML catalog into LibLouisUTDML.
   */
  public native void loadXMLCatalog(String filename);
/** Make a braille translation of the UTF-8 characters in inbuf 
according to configuration settings to UTF-8 characters in outbuf. The 
latter will be in Unicode braille. The return value is true if the 
translation is successful and false if not. Any errors are recorded in 
logFile. If this is null, they are printed on stderr. There may be 
errors even if the return value is true. If there are none, the log file 
will be empty. settingsstring may be used to pass in configuration 
settings. */

  public native boolean translateString (String configFileList,
                                         byte[]inbuf,
                                         byte[]outbuf, int[]outlen,
                                         String logFilename,
                                         String settingsString, int mode);
  
  /** the brf characters in inbuf are translated to print characters in 
      outbuf according to the settings in the configuration files and 
      setingsStrring. The translation will be in UTF-8.
  */
  
  public native boolean backTranslateString (String configFileList,
                                             byte[]inbuf,
                                             byte[]outbuf, 
                                             int[]outlen,
                                             String logFilename,
                                             String settingsSrting,
                                             int mode);
  
  /** The xml document in inputFile is translated into braille and the 
      translation is placed 
      in outputFile. The return value and log file are as described above.
  */
  
  public native boolean translateFile (String configFileList, 
                                       String inputFileName,
                                       String outputFileName,
                                       String logFileName,
                                       String settingsString, int mode);
  
  /** The plain-text file in inFile is translated to braille and the 
      translation placed in outputFile as described for the previous method. 
      If the text contains blank lines they are treated as paragraph breaks.
  */

  public native boolean translateTextFile (String configFileList,
					       String inputFileName,
					       String outputFileName,
					       String logFileName,
					       String settingsString,
					       int mode);

/* Back-translate the brf file in inFile into braille in outFile 
according to configuration specifications.
*/

  public native boolean backTranslateFile (String configFileList,
					       String inputFileName,
					       String outputFileName,
					       String logFileName,
					       String settingsString,
					       int mode);

/** Convert the utf8 character string in inbuf to Unicode braille dot 
patterns and place the result as a utf8 string in outbuf. */

public native boolean
charToDots (String tableList, byte[]inbuf,
		byte[]outbuf, String logFile,
		int mode);

/** Convert the utf8 string of dot patterns in inbuf to characters and 
place the result as a utf8 string in outbuf. */

public native boolean
dotsToChar (String tableList, byte[]inbuf,
		byte[]outbuf, String logFile,
		int mode);

/** See if the table in tableList exists and is valid. If no errors are 
found logFile will be empty. */

public native boolean
checkTable (String tableList, String logFile, int mode);

/**
* Path on which liblouis tables and liblouisutdml files can be found. 
*/

public void setDataPath(String path)
{
    LibLouis.getInstance().setDataPath(path);
}

/**
* This method performs the functions of the file2brl program in a more 
* contrrolled environment.
*/

public native boolean file2brl (String[] args);

/**
* Set the path to which temporary files will be written.
*/
public native void setWriteablePath (String path);

/** You must call free at the end of your application to free all 
memory used by liblouisutdml and liblouis. Do NOT call it after every 
call to a liblouisutdml method. This will result in great 
inefficiency. The memory used by liblouisutdml for each document is 
freed wen the method completes, but some memory holding configuration 
settings is held. The memory used by liblouis is freed only when the 
free method is called.
*/

  public native void free ();

public native void setLogFile (String fileName);

public native void logEnd();

public native void registerLogCallback(LogCallback cb);

public native void setLogLevel(int level);
}
